services/github/github-api-provider.js

```diff
@@ -52,33 +71,70 @@ class GithubApiProvider {
     if (this.withPooling) {
       this.standardTokens.add(tokenString)
       this.searchTokens.add(tokenString)
+      this.graphqlTokens.add(tokenString)
     } else {
       throw Error('When not using a token pool, do not provide tokens')
     }
   }
 
-  updateToken(token, headers) {
+  getRateLimitFromHeaders(headers) {
+    const h = Joi.attempt(headers, headerSchema)
+    return {
+      rateLimit: h['x-ratelimit-limit'],
+      totalUsesRemaining: h['x-ratelimit-remaining'],
+      nextReset: h['x-ratelimit-reset'],
+    }
+  }
+
+  getRateLimitFromBody(body) {
+    const parsedBody = JSON.parse(body)
```

Comment:
Hmm, it would be nice to avoid parsing the body twice for all these responses. Any idea if there's a good way to do that?

Reply:
Yeah this is a bit annoying.. I think we could, but I think its also a fairly major change. At the moment we're essentially doing all this token work inside the call to `_request()`, which means we can deal with manipulating the request and processing the response in the same place. We can also deal with `standardTokens`, `searchTokens` and `graphqlTokens` in the same `GithubApiProvider` class. In order to avoid double-parsing, we'd need to restructure so we can still manipulate the request, but then update the token pool after we call `_parseJson()` in either `_requestJson()` or `_requestGraphql()`.

I've not really completely thought through the consequences of all this, but here's a couple of thoughts:

One way to deal with that would be to split this up so we deal with `standardTokens` / `searchTokens` in by overriding the `_requestJson()` method in `GithubAuthV3Service`.. and `graphqlTokens` by overriding the `_requestGraphql()` method of `GithubAuthV4Service`. Although at that point if `GithubAuthV4Service` overrides `_requestGraphql()`, its not really a subclass of `BaseGraphqlService` anymore. 

Maybe another way would be to allow the `_requestJson()`/`_requestGraphql()`/`_requestFomat()` methods in base classes to accept an optional callback param or something that cna be used to run a post-process after parsing but before we return the response.

Any further thoughts or ideas?

Reply:
How about having GithubAuthV4Service rely on BaseGraphqlService to do the json parsing, but pass in an empty schema so the validation is a no-op. Then the token-related stuff can happen, then GithubAuthV4Service can invoke _validate?

----

Partly related, I dislike deep inheritance because the indirection makes it hard for people unfamiliar with the code to dig through. Each layer is one more thing to keep in your head at once.

So I think it would be good to make BaseGraphqlService extend directly from BaseService. All that would take is exporting `parseJson` as its own function (or moving it to its own file).

There are a couple places in GitHub where we invoke `_parseJson` from within a service, and those will be made cleaner by importing it directly.

In Gerritt there's also a place where we override it, though it'd be less magical and easier to understand if we used BaseService directly (invoking _request, doing the trim, then invoking parse and validate).

The refactor is a bit involved so maybe it's best to leave it for a follow-on.

Reply:
As a first step on this, I've moved `parseJson()` out to a helper and switched `BaseGraphqlService` to inherit from `BaseService`. That seems sensible anyway, but I've not taken it further.

Not sure I agree that overriding a method from the base class is "magical" in the context of gerrit though. Its what pretty much every other function we declare in a badge service class is doing - seems like a pretty consistent pattern :man_shrugging: